### ISSUE TITLE
way to open modals programmatically

### ISSUE THREAD

**Issue by drma-dev** [Megabit/Blazorise#2986]
something that doesn't need to explicitly put the modal on the page and thus refer only to the class and its parameters.

example:
https://www.matblazor.com/DialogService

**Comment by stsrki** [comment]
We already have that and it is named `IMessageService`.

https://blazorise.com/docs/services/message/
https://bootstrapdemo.blazorise.com/tests/utilities/services

**Comment by drma-dev** [comment]
I couldn't identify how to make this call.

This service only calls Predefined Modal, not one made by me.

**Comment by stsrki** [comment]
Ah, it didn't show me the code last time I opened your link on MatBlazor. You're right it is not the same thing as we have in Blazorise.

To tell you the truth I really don't see a reason for that kind of component. Seems like a regular custom Modal wrapper component with extra steps. If I were you, I would just create a component that would hold a Modal and then have methods like `Show()`, `Confirm()`, `Ask()`, etc, based on your needs.

Most of these we already have, except for Prompt with an input.

**Comment by drma-dev** [comment]
I already do these suggestions you gave me. but let me give you two examples that can justify this type of functionality:

1 - a page can have several modals (maybe even dozens). this makes the page layout a mess with so much. plus having to load all this code on every page load. having a service like this I would just open what the user requested.

2 - I have a page that has a model A, and this modal A opens a model B. If from model B I want to open the A again, I can't, because the blazor understands this as a cyclic reference. so if I open this modal by code, the blazor won't have to complain.

**Comment by drma-dev** [comment]
@stsrki  makes sense?

**Comment by stsrki** [comment]
I'm not sure why Blazor would complain about cyclic reference. It never occurred to me and I did some very complicated nesting of components in the past.

To your question. It does make sense a little. Although the thing that most bothers me is that it still requires you to write a lot of manual code. That's why I said it doesn't give much improvement. You're still writing the custom wrapper component, and you still have to reference it in some way.

I will add it to the backlog for future milestones.

**Comment by David-Moreira** [comment]
@stsrki 
This is kinda of a new component / api that needs to invoke / render the modal content programatically. Need more time to allocate. Delay to 1.1?
This one's actually a pretty good example of the feature being requested here:
https://github.com/Blazored/Modal

**Comment by stsrki** [comment]
Yeah, not enough time to properly implement it and do the testing. Going to v1.1.

**Comment by David-Moreira** [comment]
Trying to figure out if totally new component like blazored one or not. Maybe we can mostly reuse ours.

What if we have: 
App.razor
```
<Blazorise.ThemeProvider Theme="@theme">
    <Router AppAssembly="typeof(App).Assembly">
        <Found Context="routeData">
            <RouteView RouteData="routeData" DefaultLayout="typeof(MainLayout)" />
        </Found>
        <NotFound>
            <p>Sorry, there's nothing at this address.</p>
        </NotFound>
    </Router>
    <MessageAlert />
    <NotificationAlert />
    <PageProgressAlert />
    <Modal name="xpto">

    </Modal>
</Blazorise.ThemeProvider>
```
and then we have some `ModalService` we can inject into our components and
- call ModalService.Show\<MyCustomComponent>("xpto")
  - Where `MyCustomComponent` can basically have any content the user would like to place in the `Modal`, which should probably still abide to the `<ModalHeader> <ModalBody>` 
  - And here we can start borrowing some ideas from Blazored, like having a way to pass parameters to `MyCustomComponent `which is pretty good.
- call ModalService.Show("xpto")
  - This one ends up being neat too, as it lets the user define modals in a centralized place, and then just call trough the ModalService whenever
  - It will just render the currently defined modal child content, just as it works now.  



And this way we should be able to keep most of our core implementation.

**Comment by stsrki** [comment]
I don't like having `<Modal name="xpto">` in root and calling it by name with name. I think it is better to have a completely new component specialized only for showing the requested modal or other components based on type.

```diff
<Blazorise.ThemeProvider Theme="@theme">
    <Router AppAssembly="typeof(App).Assembly">
        <Found Context="routeData">
            <RouteView RouteData="routeData" DefaultLayout="typeof(MainLayout)" />
        </Found>
        <NotFound>
            <p>Sorry, there's nothing at this address.</p>
        </NotFound>
    </Router>
    <MessageAlert />
    <NotificationAlert />
    <PageProgressAlert />
+    <ModalProvider />
</Blazorise.ThemeProvider>
```

And then through modal service we can show it

`ModalService.Show( typeof( MyCustomModal ) );`

**Comment by drma-dev** [comment]
@David-Moreira I was following your conversation about this feature. Do you have an estimate of when version 1.1 will be released? was it also seen if the cyclic reference problem was solved?

**Comment by David-Moreira** [comment]
I think it is scheduled for the beginning of September.
Right you should be able. To. Open several modals.

You can test our master branch if you have the time.
@stsrki that prerelease version is it out yet?

**Comment by drma-dev** [comment]
@David-Moreira 

I didn't mean to open several modals, but to open in a cross way. blazor currently does not allow it because an error occurs.
`2 - I have a page that has a modal A, and this modal A opens a modal B. If from modal B I want to open the A again, I can't, because the blazor understands this as a cyclic reference. so if I open this modal by code, the blazor will have no reason to complain.`

https://github.com/dotnet/runtime/issues/60512

**Comment by stsrki** [comment]
> @stsrki that prerelease version is it out yet?

Yes, the 1.1 preview is already out.

**Comment by David-Moreira** [comment]
@drma-dev we would really appreciate it if you could test it with the 1.1 preview. Let us know if it does not work with a repro please.

**Comment by drma-dev** [comment]
```
blazor.webassembly.js:1 System.AggregateException: One or more errors occurred. (Type Blazorise.Bootstrap.BootstrapClassProvider has invalid vtable method slot 125 with method Blazorise.ClassProvider:DropdownObserverShow ())
 ---> System.TypeLoadException: Type Blazorise.Bootstrap.BootstrapClassProvider has invalid vtable method slot 125 with method Blazorise.ClassProvider:DropdownObserverShow ()
   at Program.<Main>$(String[] args) in C:\Users\dhiog\source\repos\drma-dev\VerusDate\src\VerusDate.Web\Program.cs:line 14
   --- End of inner exception stack trace ---
```
I can't make it work

![image](https://user-images.githubusercontent.com/12803050/184497966-3ba03f63-85c3-4435-af5d-a0fd6bd08a20.png)
maybe because the versions are different

**Comment by David-Moreira** [comment]
Strange error haha version specifically shouldn't matter if the software is compatible. But I guess it's not. 
@stsrki did you release a two Pre releases versions? One being more recent?

**Comment by David-Moreira** [comment]
Ahhh never mind I just noticed every provider preview is for v1.0. Certainly it may not be 100% compatible. @stsrki forgot to release providers for 1.1?

**Comment by stsrki** [comment]
That's strange. I can only guess that something went wrong with the build/publish process.

**Comment by stsrki** [comment]
I have fixed the NuGet push batch file and released a 1.1 preview 2. Can you try again?

**Comment by drma-dev** [comment]
@David-Moreira how do i use it? Do you already have documentation for this?

**Comment by David-Moreira** [comment]
1- You're right, no idea why... it kinda looks like it's disabled even. We'll take a look.

2-
If you set `UseModalStructure` to false then you can just straight up provide your own Modal structure. By using the `ModalHeader` / `ModalBody` / `ModalFooter` freely.
![image](https://user-images.githubusercontent.com/22283161/184996391-0f79b9fd-6eca-4e2a-9ac1-25c8d948b3ae.png)
THere might be some very specific attributes that are missing because they are set on a specific component, like the `ModalBody`. `MaxHeight` however seems worth providing as a global ModalProvider option and still propagate there? @stsrki ?

**Comment by David-Moreira** [comment]
PS: Glad it's working. If you have any futher feedback before the release, we appreciate it. :)
